torch/optim/ema_utils.py

```diff
@@ -0,0 +1,198 @@
+        parameters
+
+    Example:
+        model = Model().to(device)
+        opt = torch.optim.Adam(model.parameters())
+
+        opt = EMAOptimizer(opt, device, 0.9999)
+
+        for epoch in range(epochs):
+            training_loop(model, opt)
+
+            regular_eval_accuracy = evaluate(model)
+
+            with opt.swap_ema_weights():
+                ema_eval_accuracy = evaluate(model)
+    """
+
+    def __init__(
+        self,
+        optimizer: torch.optim.Optimizer,
+        device: torch.device,
+        decay: float = 0.9999,
+    ):
+        self.optimizer = optimizer
+        self.decay = decay
+        self.device = device
+
+        self.first_iteration = True
+        self.rebuild_ema_params = True
+        self.stream = None
+        self.thread = None
+
+        self.ema_params = ()
+
+    def all_parameters(self):
+        return (
+            param for group in self.param_groups for param in group['params']
+        )
+
+    def step(self, closure=None):
+        self.join()
+
+        if self.first_iteration:
+            if any(p.is_cuda for p in self.all_parameters()):
+                self.stream = torch.cuda.Stream()
+
+            self.first_iteration = False
+
+        if self.rebuild_ema_params:
+            opt_params = list(self.all_parameters())
+
+            self.ema_params += tuple(
+                copy.deepcopy(param.data.detach()).to(self.device)
+                for param in opt_params[len(self.ema_params):]
+            )
+            self.rebuild_ema_params = False
+
+        loss = self.optimizer.step(closure)
+
+        self.update()
+        return loss
+
+    @torch.no_grad()
+    def update(self):
+        if self.stream is not None:
+            self.stream.wait_stream(torch.cuda.current_stream())
+
+        with torch.cuda.stream(self.stream):
+            current_model_state = tuple(
+                param.data.to(self.device, non_blocking=True)
+                for param in self.all_parameters()
+            )
+
+            if self.device.type == 'cuda':
+                ema_update(self.ema_params, current_model_state, self.decay)
+
+        if self.device.type == 'cpu':
+            self.thread = threading.Thread(
+                target=run_ema_update_cpu,
+                args=(
+                    self.ema_params,
+                    current_model_state,
+                    self.decay,
+                    self.stream,
+                ),
+            )
+            self.thread.start()
+
+    @contextlib.contextmanager
+    def swap_ema_weights(self, enabled: bool = True):
+        r"""
+        A context manager to in-place swap regular parameters with EMA
+        parameters.
+        It swaps back to the original regular parameters on context manager
+        exit.
+
+        Args:
+            enabled (bool): whether the swap should be performed
+        """
+
+        def swap_tensors(tensor1, tensor2):
+            tmp = torch.empty_like(tensor1)
```

Comment:
btw this might be done via simple swapping of storage() references? (and this would incur no copy / memory accesses?) I also wonder if there is an existing util function for this - seems a useful idiom.

If actual copies are important, there is also a related request on:
- https://github.com/pytorch/pytorch/issues/90539

Reply:
Swapping references seems a bit hacky to me. It might break code that relies on storage references and I'm not aware of any other pytorch op that manipulates storage references like that. But that's up for discussion.

This part is also not meant to be called very often (only before/after validation or saving weights), so the performance is not that critical. Once a `_foreach_copy` exists we could use it of course.